### PR TITLE
Avoid non-lazy reading

### PR DESCRIPTION
**Pull request by keflavich**
This is another independent attempt to solve #774/775

### CODE REVIEW

**Comment by keflavich** [comment]
`floordiv` was incorrectly being supported in some cases; it shouldn't work!
```python
In [4]: np.array([5])//(2*u.K)
Traceback (most recent call last):
  File "<ipython-input-4-2cff403b3048>", line 1, in <module>
    np.array([5])//(2*u.K)
  File "/orange/adamginsburg/miniconda3/envs/python39/lib/python3.9/site-packages/astropy/units/quantity.py", line 466, in __array_ufunc__
    converters, unit = converters_and_unit(function, method, *inputs)
  File "/orange/adamginsburg/miniconda3/envs/python39/lib/python3.9/site-packages/astropy/units/quantity_helper/converters.py", line 192, in converters_and_unit
    raise UnitConversionError(
UnitConversionError: Can only apply 'floor_divide' function to dimensionless quantities when other argument is not a quantity (unless the latter is all zero/infinity/nan)
```

**Comment by keflavich** [comment]
ah shoot no it _should_ be supported.  I have to fix it.

**Comment by keflavich** [comment]
I think this solves a big part of #774/#775 now.  None of the tests explicitly test for data loading - I don't think I know how to do that right now - but they circumvent some of the more problematic cases of `u.Quantity(self)`

File: spectral_cube/spectral_cube.py [hunk 1]
```diff
@@ -903,30 +901,52 @@ def _get_flat_shape(self, axis):
...
         except Exception as ex:
             raise AssertionError("Function could not be applied to a simple "
                                  "cube.  The error was: {0}".format(ex))
 
-        data = function(u.Quantity(self._get_filled_data(fill=self._fill_value),
-                                   self.unit, copy=False),
-                        *args)
+        # We don't need to convert to a quantity here because the shape check
+        data_in = self._get_filled_data(fill=self._fill_value)
+        data = function(data_in, *args)
```
Comment (keflavich) [inline spectral_cube/spectral_cube.py:942]: do we need to explicitly delete `data_in` here?

File: spectral_cube/tests/test_dask.py [hunk 2]
```diff
@@ -239,6 +241,15 @@ def func(x, add=None):
                                    rslt3.filled_data[:].value)
 
 
+def test_cube_on_cube(data_adv):
+    # regression test for #782
+    cube = DaskSpectralCube.read(data_adv)
+
+    with patch.object(cube, '_cube_on_cube_operation') as mock:
+        cube * cube
+    mock.assert_called_once()
```
Comment (keflavich) [inline spectral_cube/tests/test_dask.py:250]: maybe this is where we need an explicit deletion?
```suggestion
    mock.assert_called_once()
    del cube
```

File: spectral_cube/tests/test_spectral_cube.py [hunk 3]
```diff
@@ -344,6 +342,36 @@ def test_apply_everywhere(self, operation, value, data_advs, use_dask):
...
+        assert c1.unit == value.unit
+
+        # append 'o' to indicate that it has been operated on
+        # value.value: the __add__ function explicitly drops the units
+        c1o = c1._apply_everywhere(operation, value.value, check_units=False)
+        d1o = operation(u.Quantity(d1, u.K), value)
+        assert c1o.unit == c1.unit
+        assert c1o.unit == value.unit
+
+        assert np.all(d1o == c1o.filled_data[:])
```
Comment (keflavich) [inline spectral_cube/tests/test_spectral_cube.py:360]: not needed above, but here?
```suggestion
        assert np.all(d1o == c1o.filled_data[:])
        
        del c1
        del c1o
```

File: spectral_cube/tests/test_spectral_cube.py [hunk 4]
```diff
@@ -344,6 +342,36 @@ def test_apply_everywhere(self, operation, value, data_advs, use_dask):
...
+
+    # The workaround for dask data-loading prevents this test from succeeding
+    @pytest.mark.xfail(raises=u.UnitConversionError, strict=True)
+    @pytest.mark.parametrize(('operation', 'value'),
+                             ((operator.div if hasattr(operator,'div') else operator.floordiv, 0.5*u.K),))
+    def test_apply_everywhere_floordivide(self, operation, value, data_advs, use_dask):
+        c1, d1 = cube_and_raw(data_advs, use_dask=use_dask)
+
+        # append 'o' to indicate that it has been operated on
+        c1o = c1._apply_everywhere(operation, value)
```
Comment (keflavich) [inline spectral_cube/tests/test_spectral_cube.py:370]: if this is causing a teardown error, there's nothing we can do about it - this is an xfail

File: spectral_cube/tests/test_spectral_cube.py [hunk 5]
```diff
@@ -344,6 +342,36 @@ def test_apply_everywhere(self, operation, value, data_advs, use_dask):
...
+        d1o = operation(u.Quantity(d1, u.K), value)
+
+        assert np.all(d1o == c1o.filled_data[:])
```
Comment (keflavich) [inline spectral_cube/tests/test_spectral_cube.py:373]: Delete these lines as redundant; the exception should be raised in the previous line
```suggestion
```

**Review by keflavich — COMMENTED** [review]
I'm trying to track down where the open file is coming from

**Comment by keflavich** [comment]
I had to totally remove the "expected failure" test for floordiv because it was leaving files open - I don't understand that at all.

Otherwise, this goes a long way to avoiding the early reads of full cubes

File: spectral_cube/tests/test_spectral_cube.py [hunk 6]
```diff
@@ -344,6 +342,40 @@ def test_apply_everywhere(self, operation, value, data_advs, use_dask):
...
+    @pytest.mark.parametrize(('operation', 'value'), ((operator.add, 0.5*u.K),
+                                                      (operator.sub, 0.5*u.K),))
+    def test_apply_everywhere_plusminus(self, operation, value, data_advs, use_dask):
+        c1, d1 = cube_and_raw(data_advs, use_dask=use_dask)
+
+        assert c1.unit == value.unit
+
+        # append 'o' to indicate that it has been operated on
+        # value.value: the __add__ function explicitly drops the units
+        c1o = c1._apply_everywhere(operation, value.value, check_units=False)
```
Comment (e-koch) [inline spectral_cube/tests/test_spectral_cube.py:355]: Add a test case with `check_units=True`?

**Review by e-koch — APPROVED** [review]
The code looks good to me, though I don't have a good grasp of the `u.Quantity` issue and I think should wait for @astrofrog's review.

**Comment by e-koch** [comment]
Is #782 needed anymore? It looks like all of those changes are here.

**Comment by keflavich** [comment]
This is branched off of #782; that should be merged first

File: spectral_cube/tests/test_spectral_cube.py (hunk 6 above)
Comment (keflavich) [reply spectral_cube/tests/test_spectral_cube.py:355]: `add` is only supposed to use `check_units=False`; multiply/divide use `check_units=True`

**Comment by keflavich** [comment]
The commit history on this is a bit messed up right now; it's overlapped with #782 , which I'd like to resolve first; I'll have to rebase against that

File: spectral_cube/tests/test_dask.py [hunk 7]
```diff
@@ -239,6 +241,40 @@ def func(x, add=None):
...
+    assert isinstance(cube, (DaskSpectralCube, DaskVaryingResolutionSpectralCube))
+    cube2 = SpectralCube.read(dataname, use_dask=False)
+    if 'image' not in filename:
+        # 'image' would be CASA and must be dask
+        assert not isinstance(cube2, (DaskSpectralCube, DaskVaryingResolutionSpectralCube))
+
+    with patch.object(cube, '_cube_on_cube_operation') as mock:
+        cube * cube
+    mock.assert_called_once()
+    del cube
```
Comment (dhomeier) [inline spectral_cube/tests/test_dask.py:264]:
```suggestion
```
Well, this must certainly not be deleted before you're done with it!

**Review by dhomeier — COMMENTED** [review]
I ran into some additional `open_files` failures trying to debug this locally, notably in the `oned_`, `twod_` tests the skipped runs were somehow interfering with closing the file; OTOH some of the other explicit removals were no longer necessary and I also got `test_apply_everywhere_floordivide` to pass. The changes are a bit lengthy; I can open another PR against this branch, but if this passes on the CI as is, maybe it's not critical.

**Pull request merged**